scripts/asr_language_modeling/ngram_lm/eval_beamsearch_ngram_transducer.py

```diff
@@ -13,257 +13,420 @@
+from typing import List, Optional
 
 import editdistance
+import numpy as np
 import torch
-from omegaconf import OmegaConf
+from omegaconf import MISSING, OmegaConf
+from sklearn.model_selection import ParameterGrid
 from tqdm.auto import tqdm
 
 import nemo.collections.asr as nemo_asr
-from nemo.collections.asr.metrics.rnnt_wer_bpe import RNNTBPEDecodingConfig
+from nemo.collections.asr.parts.submodules import rnnt_beam_decoding
+from nemo.core.config import hydra_runner
 from nemo.utils import logging
 
+# fmt: off
+
+
+@dataclass
+class EvalBeamSearchNGramConfig:
+    """
+    Evaluate an ASR model with beam search decoding and n-gram KenLM language model.
+    """
+    # # The path of the '.nemo' file of the ASR model or the name of a pretrained model (ngc / huggingface)
+    nemo_model_file: str = MISSING
+
+    # File paths
+    input_manifest: str = MISSING  # The manifest file of the evaluation set
+    kenlm_model_file: Optional[str] = None  # The path of the KenLM binary model file
+    preds_output_folder: Optional[str] = None  # The optional folder where the predictions are stored
+    probs_cache_file: Optional[str] = None  # The cache file for storing the logprobs of the model
+
+    # Parameters for inference
+    acoustic_batch_size: int = 128  # The batch size to calculate log probabilities
+    beam_batch_size: int = 128  # The batch size to be used for beam search decoding
+    device: str = "cuda"  # The device to load the model onto to calculate log probabilities
+    use_amp: bool = False  # Whether to use AMP if available to calculate log probabilities
+
+    # The decoding scheme to be used for evaluation
+    decoding_strategy: str = "greedy_batch" # ["greedy_batch", "maes"]
+
+    # Beam Search hyperparameters
+    beam_width: List[int] = field(default_factory=lambda: [128])  # The width or list of the widths for the beam search decoding
+    beam_alpha: List[float] = field(default_factory=lambda: [0.2])  # The alpha parameter or list of the alphas for the beam search decoding
+
+    maes_prefix_alpha: List[int] = field(default_factory=lambda: [2])  # The maes_prefix_alpha or list of the maes_prefix_alpha for the maes decoding
+    maes_expansion_gamma: List[float] = field(default_factory=lambda: [2.3])  # The maes_expansion_gamma or list of the maes_expansion_gamma for the maes decoding
+
+    # HAT related parameters (only for internal lm subtraction)
+    hat_subtract_ilm: bool = False
+    hat_ilm_weight: List[float] = field(default_factory=lambda: [0.0])
+
+    decoding: rnnt_beam_decoding.BeamRNNTInferConfig = rnnt_beam_decoding.BeamRNNTInferConfig(beam_size=128)
+
+
+# fmt: on
+
+
+def decoding_step(
+    model: nemo_asr.models.ASRModel,
+    cfg: EvalBeamSearchNGramConfig,
+    all_probs: List[torch.Tensor],
+    target_transcripts: List[str],
+    preds_output_file: str = None,
+    lm_path: str = None,
+    beam_alpha: float = 0.2,
+    beam_width: int = 128,
+    maes_prefix_alpha: int = 2,
```

Comment:
Revert all these additional arguments and pass full rnntbeamdecodingconfig here.

Reply:
Done